testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_transactions.py

```diff
@@ -0,0 +1,420 @@
+from keywords import couchbaseserver, document
+from CBLClient.Document import Document
+from concurrent.futures import ThreadPoolExecutor
+import pytest
+
+
+@pytest.mark.transactions
+@pytest.mark.syncgateway
+@pytest.mark.cbl
+def test_transactions_insert_replace_remove_rollback(params_from_base_test_setup):
+    '''
+    @summary:
+    https://docs.google.com/spreadsheets/d/1YWP_iwYz9B3ej_ksLf-XKsLj4RnbuSigf7hp0BPYv0w/edit#gid=0
+    Covered #7, #8, #10, #12
+    1. Insert a transaction with 3 or more docs
+    2. replication from server to SGW and to CBL
+    3. Verify all docs in transactions are pulled to CBL
+    4. replace the transaction with new content
+    5. Verify doc is replicated to cbl whiich commited in transaction
+    6. Delete the transaction
+    7. Verify doc is removed in cbl
+    8. Rollback the transaction
+    9. Verify rollback docs are not replicated to cbl
+    '''
+
+    xattrs_enabled = params_from_base_test_setup['xattrs_enabled']
+    if not xattrs_enabled:
+        pytest.skip('this test require --xattrs flag')
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    base_url = params_from_base_test_setup["base_url"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    sg_admin_url = params_from_base_test_setup["sg_admin_url"]
+    cbl_db = params_from_base_test_setup["source_db"]
+    sg_blip_url = params_from_base_test_setup["target_url"]
+    db = params_from_base_test_setup["db"]
+    cluster_helper = ClusterKeywords(cluster_config)
+    topology = cluster_helper.get_cluster_topology(cluster_config)
+    cluster_servers = topology["couchbase_servers"]
+    cbs_one_url = cluster_servers[0]
+    cb_server = couchbaseserver.CouchbaseServer(cbs_one_url)
+
+    sg_db = "db"
+    username = "autotest"
+    password = "password"
+    channels = ["TRANSACTIONS"]
+    num_of_docs = 4
+    doc_id = "transaction-sync-id"
+    doc_ids = ""
+
+    sg_client = MobileRestClient()
+    doc_obj = Document(base_url)
+    # Reset cluster to ensure no data in system
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+    bucket = cb_server.get_bucket_names()[0]
+
+    # disable replicas and rebalance the server
+    cb_server.disable_replicas(bucket)
+    cb_server.rebalance_server(cluster_servers)
+
+    # randomly  generate replication ids
+    for i in range(0, num_of_docs):
+        doc_id = "txn_id_" + str(time.time())
+        doc_ids = doc_id + "," + doc_ids
+    # 1. Insert a transaction with 3 or more docs
+    opt_doc_ids = doc_ids.replace("txn_id_", "insert-txn_id_")
+    transactions_app_dir = "resources/data/transaction-performer.jar"
+    doc_body = document.create_doc(doc_id=doc_id, content="doc1", channels=channels)
+    data = json.dumps(doc_body)
+    cmd = "java -cp {} com.couchbase.transaction clusterip={} operationdocids={} bucket={} doccontent='{}'".format(transactions_app_dir, c.servers[0].host, opt_doc_ids, bucket, data)
```

Comment:
could you add a little bit more explanation here? you pass 4 comma divided doc ids to the jar but only 1 single data value, (content is doc1), what the jar will do? apply the data value to all 4 doc ids on sgw?

Reply:
done